test/jdk/java/nio/channels/DatagramChannel/SelectWhenRefused.java

```diff
@@ -94,9 +128,32 @@ public static void main(String[] args) throws IOException {
     }
 
     static void sendDatagram(DatagramChannel dc, SocketAddress remote)
-        throws IOException
-    {
-        ByteBuffer bb = ByteBuffer.wrap("Greetings!".getBytes());
+            throws IOException {
+        ByteBuffer bb = ByteBuffer.wrap(GREETINGS_MESSAGE.getBytes());
         dc.send(bb, remote);
     }
+
+    static boolean checkUnexpectedWakeup(DatagramChannel dc, Set<SelectionKey> selectedKeys) {
```

Comment:
For the next iteration, would you mind dropping the selectedKeys parameter from this method. It doesn't make sense to be calling dc.receive in this loop. It would be a lot simpler to do the non-blocking dc.receive and print the sender/details when not null.

Reply:
It make no sense because it puts the dc.receive into the loop that is examining the keys. If you really want to look at the state of the key then capture the result of the register method so you can print it out after the selection operation.

Reply:
I see. So you advise to pass the single key we obtained when registering the channel to the method rather than the selected key set. That's OK with me too.

Reply:
The method need only be passed one parameter, the DatagramChannel or the SelectionKey (key::channel will give you the channel). Right now it is confused because it wants to loop over all selected keys but it doesn't use the key's channel for the receive method. If I were changing this test then I think I would just pass it the DatagramChannel as the selection key is not really interesting. I think I would also change it to check if sa (the sender) is null so that it doesn't print a confusing message when no datagram has been received.

Reply:
I had a suspicion that maybe the selector could be woken up if the key became invalid - even though since this is a datagram channel and we're not closing it, I don't see how it could happen. If the key is invalid, there would be nothing to read. In addition, `n` is not supposed to be `> 0`, therefore something unexpected is going on. It seems useful to double check our assumptions: n > 0 because the key is active for reading. I'd like to check that. Hence the idea that passing the key was important. And you only want to attempt reading if the key is active for reading. Otherwise we'd get some other kind of exception and we would still not be the wiser. Good point about calling `key::channel` though!

Reply:
So remove all the key prints and just attempt dc.receive and try to show its result?

Reply:
Please no. Pass the key (or selected key set) but not the dc, and obtain the dc by calling key::channel.

Reply:
sorry your last comment hadn't loaded in for me on the github page for some reason, I hadn't been responding to it

Seeing it now though it sounds good